src/thread_event.c

```diff
@@ -0,0 +1,252 @@
+		 * thread_event_update() call later to properly set the	\
+		 * next next_event; if there are multiple events, then	\
+		 * here we can get the minimum remaining wait time to	\
+		 * the next already set event.				\
+		 */							\
+		if (event_wait > 0 && event_wait < wait) {		\
+			wait = event_wait;				\
+		}							\
+	}
+
+	ITERATE_OVER_ALL_WAITS
+#undef W
+
+	/*
+	 * In case none of the events has its condition on, set next_event to
+	 * be THREAD_ALLOCATED_NEXT_EVENT_FAST_MAX to minimize malloc() slow
+	 * path triggering.
+	 */
+	if (no_event_on) {
+		/* The subtraction is intentionally susceptible to underflow. */
+		wait = THREAD_ALLOCATED_NEXT_EVENT_FAST_MAX -
+		    thread_allocated_after;
+		if (wait == 0U) {
+			/*
+			 * We don't allow zero wait time, so delay to the next
+			 * allocation call.  The next call is guaranteed to
+			 * trigger an event and at that time
+			 * thread_allocated_after is guaranteed to be different
+			 * from THREAD_ALLOCATED_NEXT_EVENT_FAST_MAX.
+			 */
+			wait = 1U;
+		}
+	}
+
+	thread_allocated_last_event_set(tsd, thread_allocated_after);
+	thread_event_adjust_thresholds_helper(tsd,
+	    thread_allocated_after + wait);
+
+	thread_event_assert_invariants(tsd);
+}
+
+void
+thread_event_rollback(tsd_t *tsd, size_t diff) {
+	thread_event_assert_invariants(tsd);
+
+	if (diff == 0) {
+		return;
+	}
+
+	uint64_t thread_allocated = thread_allocated_get(tsd);
+
+	/* The subtraction is intentionally susceptible to underflow. */
+	uint64_t thread_allocated_rollback = thread_allocated - diff;
+
+	uint64_t last_event = thread_allocated_last_event_get(tsd);
+
+	/* Both subtractions are intentionally susceptible to underflow. */
+	if (thread_allocated_rollback - last_event <=
+	    thread_allocated - last_event) {
+		thread_allocated_set(tsd, thread_allocated_rollback);
+		thread_event_assert_invariants(tsd);
+		return;
+	}
+
+	/* The subtraction is intentionally susceptible to underflow. */
+	uint64_t wait_diff = last_event - thread_allocated_rollback;
+	assert(wait_diff <= diff);
+
+	/*
+	 * We need to shift back last_event by wait_diff, and increase the wait
+	 * times by wait_diff.  However, we cannot let last_event underflow
+	 * next_event, or let the wait times overflow.  So we first compute the
+	 * required min distance between wait_diff and UINT64_MAX, and if
+	 * wait_diff is too close to UINT64_MAX, we'll reduce it so that it
+	 * meets the min distance requirement.
+	 */
+	uint64_t next_event = thread_allocated_next_event_get(tsd);
+
+	/* The subtraction is intentionally susceptible to underflow. */
+	uint64_t dist = next_event - last_event;
+
+#define W(event, condition)						\
+	if (condition) {						\
+		uint64_t event_wait = event##_event_wait_get(tsd);	\
+		if (event_wait > 0 && event_wait > dist) {		\
+			dist = event_wait;				\
+		}							\
+	}
+
+	ITERATE_OVER_ALL_WAITS
+#undef W
+
+	if (wait_diff > UINT64_MAX - dist) {
+		/* Reduce wait_diff to avoid overflow on any wait time. */
```

Comment:
This feels way too complicated. Will it work if we do, 

     if (this_event_wait_will_overflow) { new_event_wait = WAIT_MAX; }
     ##_event_wait_set(tsd, new_event_wait); 

Let's try simplify the logic here. We should be able to rollback last_event, increase wait_time, and then compute next_wait normally (using thread_event_update)?